warehouse/accounts/security_policy.py

```diff
@@ -0,0 +1,238 @@
+                BasicAuthFailedPassword(),
+                "Invalid or non-existent authentication information. "
+                "See {projecthelp} for more information.".format(
+                    projecthelp=request.help_url(_anchor="invalid-auth")
+                ),
+            )
+
+
+@implementer(ISecurityPolicy)
+class SessionSecurityPolicy:
+    def __init__(self):
+        self._session_helper = SessionAuthenticationHelper()
+        self._principals = []
+
+    def identity(self, request):
+        # If we're calling into this API on a request, then we want to register
+        # a callback which will ensure that the response varies based on the
+        # Cookie header.
+        request.add_response_callback(add_vary_callback("Cookie"))
+        request.authentication_method = AuthenticationMethod.SESSION
+
+        # NOTE: Why is this authenticated_userid? I guess it's already
+        # been validated by session validation?
+        userid = self._session_helper.authenticated_userid(request)
+
+        # Session authentication cannot be used for uploading
+        if request.matched_route.name in ["forklift.legacy.file_upload"]:
+            return None
+
+        # Sessions can only authenticate users, not any other type of identity.
+        login_service = request.find_service(IUserService, context=None)
+        return login_service.get_user(userid)
+
+    def forget(self, request, **kw):
+        return self._session_helper.forget(request, **kw)
+
+    def remember(self, request, userid, **kw):
+        return self._session_helper.remember(request, userid, **kw)
+
+
+@implementer(ISecurityPolicy)
+class BasicAuthSecurityPolicy:
+    def callback(self, userid, request):
+        if credentials := extract_http_basic_credentials(request):
+            username, password = credentials
+            return self._check(username, password, request)
+        return None
+
+    def identity(self, request):
+        # If we're calling into this API on a request, then we want to register
+        # a callback which will ensure that the response varies based on the
+        # Authorization header.
+        request.add_response_callback(add_vary_callback("Authorization"))
+        request.authentication_method = AuthenticationMethod.BASIC_AUTH
+
+        credentials = extract_http_basic_credentials(request)
+        if credentials is None:
+            return None
+
+        username, password = credentials
+        if not _basic_auth_check(username, password, request):
+            return None
+
+        # Like sessions; basic auth can only authenticate users.
+        login_service = request.find_service(IUserService, context=None)
+        return login_service.get_user_by_username(username)
+
+    def forget(self, request, **kw):
+        # No-op.
+        return []
+
+    def remember(self, request, userid, **kw):
+        # NOTE: We could make realm configurable here.
+        return [("WWW-Authenticate", 'Basic realm="Realm"')]
```

Comment:
I cribbed this directly from the previous (Pyramid 1.0-style) authn policy: we were already implicitly returning this header, so I figured we should continue to do so on the (exceedingly unlikely) chance that someone is actually expecting it.

> I'm especially doubtful given that our MultiAuthSecurityPolicy calls remember on every subpolicies when we call its remember method, so as far as I can tell, this would result in this header be sent back to the browser when we log in ? And I guess the reason the browser doesn't display an auth modal is because it's not accompanied by a 401 Unauthenticated ?

Yep, correct. The previous (`pyramid_multiauth`-based) implementation also did this, mind you: https://github.com/mozilla-services/pyramid_multiauth/blob/e3d2f5d75201b05daefd0c589f44fb55fdf8e877/pyramid_multiauth/__init__.py#L137-L146

(I think it's pretty ugly that we always return all `remember` and `forget` headers, but I'm not sure there's a straightforward way to do that while composing AuthN policies in the way we are.)